src/storage/mod.rs

```diff
@@ -4871,4 +4890,302 @@ mod tests {
+        // fail.
+        let storage = TestStorageBuilder::new().build().unwrap();
+        let (tx, rx) = channel();
+        let expect_pessimistic_txn_not_enabled_cb = || {
+            expect_fail_callback(tx.clone(), 0, |e| match e {
+                Error::PessimisticTxnNotEnabled => (),
+                e => panic!("unexpected error chain: {:?}", e),
+            })
+        };
+        storage
+            .async_acquire_pessimistic_lock(
+                Context::default(),
+                vec![],
+                b"x".to_vec(),
+                1,
+                Options::default(),
+                expect_pessimistic_txn_not_enabled_cb(),
+            )
+            .unwrap();
+        rx.recv().unwrap();
+
+        let mut options = Options::default();
+        options.for_update_ts = 2;
+        storage
+            .async_prewrite(
+                Context::default(),
+                vec![],
+                b"x".to_vec(),
+                1,
+                options,
+                expect_pessimistic_txn_not_enabled_cb(),
+            )
+            .unwrap();
+        rx.recv().unwrap();
+
+        storage
+            .async_pessimistic_rollback(
+                Context::default(),
+                vec![],
+                1,
+                2,
+                expect_pessimistic_txn_not_enabled_cb(),
+            )
+            .unwrap();
+        rx.recv().unwrap();
+    }
+
+    // The logic of pessimistic transaction is tested in storage/mvcc/txn.rs.
+    // It tests whether Storage uses functions rightly.
+    #[test]
+    fn test_pessimistic_txn() {
+        let storage = TestStorageBuilder::from_engine(TestEngineBuilder::new().build().unwrap())
+            .lock_mgr(DummyLockMgr)
+            .build()
+            .unwrap();
+        let (tx, rx) = channel();
+
+        // Normal
+        let key = Key::from_raw(b"a");
+        storage
+            .async_acquire_pessimistic_lock(
+                Context::default(),
+                vec![(key.clone(), true)],
+                key.to_raw().unwrap(),
+                10,
+                Options::default(),
+                expect_ok_callback(tx.clone(), 0),
+            )
+            .unwrap();
+        let mut options = Options::default();
+        options.for_update_ts = 10;
+        options.is_pessimistic_lock = vec![true];
+        storage
+            .async_prewrite(
+                Context::default(),
+                vec![Mutation::Put((key.clone(), b"v".to_vec()))],
+                key.to_raw().unwrap(),
+                10,
+                options,
+                expect_ok_callback(tx.clone(), 1),
+            )
+            .unwrap();
+        storage
+            .async_commit(
+                Context::default(),
+                vec![key.clone()],
+                10,
+                20,
+                expect_ok_callback(tx.clone(), 2),
+            )
+            .unwrap();
+        rx.recv().unwrap();
```

Comment:
I worry this may cause the test unstable. I think we'd better call `rx.recv()` after each command.

Reply:
Ah.. Actually it won't make it unstable (the latch keeps the order you push the task), but I think it is better to send the next command after receiving the previous one because it's the reality...